src/java.base/share/classes/java/lang/foreign/MemorySegment.java

```diff
@@ -1966,4 +1980,69 @@ private static long getBaseAndScale(Class<?> arrayType) {
             throw new IllegalArgumentException("Not a supported array class: " + arrayType.getSimpleName());
         }
     }
+
+    /**
+     * Finds and returns the relative offset, in bytes, of the first mismatch between the source and the destination
+     * segments. More specifically, the bytes at offset {@code srcOffset} through {@code srcOffset + bytes - 1} in the
+     * source segment are compared against the bytes at offset {@code dstOffset} through {@code dstOffset + bytes - 1}
+     * in the destination segment.
+     * <p>
+     * If a mismatch is detected, the returned offset is relative to the {@linkplain #address() base address} of each
+     * segment and will be in the range of 0 (inclusive) up to the {@code length}. Otherwise, there is no mismatch and
+     * {@code -1} is returned.
+     *
+     * @param srcSegment the source segment.
+     * @param srcOffset the starting offset, in bytes, of the source segment.
+     * @param dstSegment the destination segment.
+     * @param dstOffset the starting offset, in bytes, of the destination segment.
+     * @param bytes the number of bytes to be compared.
+     * @return the relative offset, in bytes, of the first mismatch between the source and destination segments,
+     * otherwise -1 if no mismatch.
+     * @throws IllegalStateException if the {@linkplain #session() session} associated with {@code srcSegment} is not
+     * {@linkplain MemorySession#isAlive() alive}.
+     * @throws WrongThreadException if this method is called from a thread other than the thread owning
+     * the {@linkplain #session() session} associated with {@code srcSegment}.
+     * @throws IllegalStateException if the {@linkplain #session() session} associated with {@code dstSegment} is not
+     * {@linkplain MemorySession#isAlive() alive}.
+     * @throws WrongThreadException if this method is called from a thread other than the thread owning
+     * the {@linkplain #session() session} associated with {@code dstSegment}.
+     * @throws IndexOutOfBoundsException if {@code srcOffset + bytes > srcSegment.byteSize()} or if
+     * {@code dstOffset + bytes > dstSegment.byteSize()}, or if either {@code srcOffset}, {@code dstOffset}
+     * or {@code bytes} are {@code < 0}.
+     *
+     * @see MemorySegment#mismatch(MemorySegment)
+     */
+    static long mismatch(MemorySegment srcSegment, long srcOffset, MemorySegment dstSegment, long dstOffset, long bytes) {
```

Comment:
Should we make the signature the same as that for Arrays.mismatch e.g.:
```
    public static int mismatch(byte[] a, int aFromIndex, int aToIndex,
                               byte[] b, int bFromIndex, int bToIndex) {
```
?
It's tempting to make both mismatch methods static.

Reply:
Re: static vs instance, apologies I don't mean to reopen that discussion.

IIRC the motivation for the signatures in `Arrays` was in part because that is the pattern used by all the other methods. You are right that the underlying primitive is passing in a common length, which is what the supporting internal mismatch methods use, but it may not be as helpful to require the user to compute the minimum length. To be honest I am torn between keeping close to the `copy` signature or using absolute ranges :-)

Reply:
One advantage to push the length computation inside is that explaining what the instance method does become trivial. An option could be to expose both, but seems like overkill. I think I'm fine with exposing the Arrays signature - after all the user can always get what he wants either way. That would at least make the API point consistent with Arrays.mismatch (in the same way MemorySegment::copy is consistent with System::arraycopy).

Reply:
I agree with that positioning, `System::arraycopy` is to `MemorySegment::copy` as `Arrays::mistmatch` is to `MemorySegment::mistmatch`.